### ISSUE TITLE
UseCase for RSU

### ISSUE THREAD

**Issue by ruipatcheco** [riebl/artery#66]
Hello,

Im trying to make an RSU DEN use case

My scenario:
A truck controlled by Storyboard will stop and send a DEN message, and RSU receiving this message will re-broadcast it for 10 minutes

My work:
I added my DEN use case (IcyRoadUseCaseRSU) under artery/src/artery/application/den/ and modified mt-its2017 scenario to use my usecase. The use case's class (IcyRoadRSU) inherits from the provided UseCase.cc (class IcyRoadRSU : public UseCase).

![2019-03-13-173618_738x450_scrot](https://user-images.githubusercontent.com/15826810/54301290-90f74780-45b6-11e9-8910-0154822c30f5.png)

The problem and relevant test:
When using the DEN use case, i get the following error:

no valid object of type 'N6artery19VehicleDataProviderE' registered -- in module (artery::den::IcyRoadRSU) World.rsu[0].middleware.DEN.IcyRoadRSU (id=155), during network initialization

My diagnosis:
It comes from the UseCase.cc being targeted at vehicles and not RSUs because it uses VehicleDataProvider

![2019-03-13-173335_544x173_scrot](https://user-images.githubusercontent.com/15826810/54301082-2b0ac000-45b6-11e9-957a-d05e415326d8.png)

This issue is very much related to https://github.com/riebl/artery/issues/39

Is there any UseCase capable of being used in RSU already implemented? If so, where/how can I find it?

Best regards and thank you very much for your work and great support of the community with excelent answers!!

**Comment by riebl** [comment]
No, at the moment Artery does not include specific DEN use cases for RSUs. Contributions are welcome :-)
The underlying `DenService` currently works only with vehicles that have a `VehicleDataProvider` through `VehicleMiddleware`. This restriction can be eliminated though as it is only used for retrieving the station ID.

**Comment by ruipatcheco** [comment]
Ok, thanks for the answer! I'll be sure to contribute when its ready

**Comment by ruipatcheco** [comment]
I would like to use Vehicle to RSU communication as described above, so I've since modified the DenService as here: https://gist.github.com/MounaKaroui/ea65ce3d69975f07f21858a5e8011ffd

to remove the dependency from VehicleDataProvider so that I can use it with an RSU.

I've created a new UseCaseRSU.cc based on  UseCase.cc where I've removed the dependency from mVdp (VehicleDataProvider).

mVdp = &mService->getFacilities().get_const<VehicleDataProvider>();


The vehicles are running IcyRoad usecase and the RSU's are running IcyRoadRSU usecase, the difference being one inherits from UseCase.cc and the other from UseCaseRSU.cc, besides that, all the code is the same.

My goal is for a DEN message sent from a vehicle to be received by the RSU's IcyRoadRSU::indicate(const artery::DenmObject& denm) method.

All is compiling and RSU's are visible in OMNeT window during simulation
![2019-03-29-161539_1283x646_scrot](https://user-images.githubusercontent.com/15826810/55246855-1547fc80-523e-11e9-930c-aab2e3975d52.png)

But, the RSU's dont seem to be doing anything because IcyRoadRSU::check() is not being invoked for each middleware time step. 

The omnet.ini file (https://pastebin.com/77UFGZzz) has *.rsu[*].middleware.updateInterval = 0.1s 


Any ideas from where the problem might be coming from?

Best regards and thank you very much for your work and great support of the community with excellent answers!!

**Comment by ruisonghan** [comment]
> Hi again,
> 
> I would like to use Vehicle to RSU communication as described above, so I've since modified the DenService as here: https://gist.github.com/MounaKaroui/ea65ce3d69975f07f21858a5e8011ffd
> 
> to remove the dependency from VehicleDataProvider so that I can use it with an RSU.
> 
> I've created a new UseCaseRSU.cc based on UseCase.cc where I've removed the dependency from mVdp (VehicleDataProvider).
> 
> mVdp = &mService->getFacilities().get_const();
> 
> The vehicles are running IcyRoad usecase and the RSU's are running IcyRoadRSU usecase, the difference being one inherits from UseCase.cc and the other from UseCaseRSU.cc, besides that, all the code is the same.
> 
> My goal is for a DEN message sent from a vehicle to be received by the RSU's IcyRoadRSU::indicate(const artery::DenmObject& denm) method.
> 
> All is compiling and RSU's are visible in OMNeT window during simulation
> ![2019-03-29-161539_1283x646_scrot](https://user-images.githubusercontent.com/15826810/55246855-1547fc80-523e-11e9-930c-aab2e3975d52.png)
> 
> But, the RSU's dont seem to be doing anything because IcyRoadRSU::check() is not being invoked for each middleware time step.
> 
> The omnet.ini file (https://pastebin.com/77UFGZzz) has _.rsu[_].middleware.updateInterval = 0.1s
> 
> Any ideas from where the problem might be coming from?
> 
> Best regards and thank you very much for your work and great support of the community with excellent answers!!

Hi,

May I know why did you change the input of ```DenService::indicate()``` ?

**Comment by riebl** [comment]
@ruipatcheco is your `IcyRoadRSU` actually instantiated at your RSUs' middleware? Can you see it in the module tree explorer?

**Comment by ruipatcheco** [comment]
> > Hi again,
> > I would like to use Vehicle to RSU communication as described above, so I've since modified the DenService as here: https://gist.github.com/MounaKaroui/ea65ce3d69975f07f21858a5e8011ffd
> > to remove the dependency from VehicleDataProvider so that I can use it with an RSU.
> > I've created a new UseCaseRSU.cc based on UseCase.cc where I've removed the dependency from mVdp (VehicleDataProvider).
> > mVdp = &mService->getFacilities().get_const();
> > The vehicles are running IcyRoad usecase and the RSU's are running IcyRoadRSU usecase, the difference being one inherits from UseCase.cc and the other from UseCaseRSU.cc, besides that, all the code is the same.
> > My goal is for a DEN message sent from a vehicle to be received by the RSU's IcyRoadRSU::indicate(const artery::DenmObject& denm) method.
> > All is compiling and RSU's are visible in OMNeT window during simulation
> > ![2019-03-29-161539_1283x646_scrot](https://user-images.githubusercontent.com/15826810/55246855-1547fc80-523e-11e9-930c-aab2e3975d52.png)
> > But, the RSU's dont seem to be doing anything because IcyRoadRSU::check() is not being invoked for each middleware time step.
> > The omnet.ini file (https://pastebin.com/77UFGZzz) has _.rsu[_].middleware.updateInterval = 0.1s
> > Any ideas from where the problem might be coming from?
> > Best regards and thank you very much for your work and great support of the community with excellent answers!!
> 
> Hi,
> 
> May I know why did you change the input of `DenService::indicate()` ?

Hi,

I don't think I did, as you can see, this is the latest version on artery's git: https://github.com/riebl/artery/blob/master/src/artery/application/DenService.cc

void DenService::indicate(const vanetza::btp::DataIndication& indication, std::unique_ptr<vanetza::UpPacket> packet)

and it is the same

**Comment by ruipatcheco** [comment]
> @ruipatcheco is your `IcyRoadRSU` actually instantiated at your RSUs' middleware? Can you see it in the module tree explorer?

Hi, 

thank you for helping me,

Yes, here it is:

![2019-04-01-180119_805x216_scrot](https://user-images.githubusercontent.com/15826810/55345610-390e7b00-54a8-11e9-980d-ce71bba38047.png)

![2019-04-01-180126_171x147_scrot](https://user-images.githubusercontent.com/15826810/55345612-3d3a9880-54a8-11e9-8c02-79966ed7006b.png)

![2019-04-01-180134_203x132_scrot](https://user-images.githubusercontent.com/15826810/55345614-40ce1f80-54a8-11e9-80ae-5e6ef7b0bf53.png)

**Comment by ruisonghan** [comment]
> > > Hi again,
> > > I would like to use Vehicle to RSU communication as described above, so I've since modified the DenService as here: https://gist.github.com/MounaKaroui/ea65ce3d69975f07f21858a5e8011ffd
> > > to remove the dependency from VehicleDataProvider so that I can use it with an RSU.
> > > I've created a new UseCaseRSU.cc based on UseCase.cc where I've removed the dependency from mVdp (VehicleDataProvider).
> > > mVdp = &mService->getFacilities().get_const();
> > > The vehicles are running IcyRoad usecase and the RSU's are running IcyRoadRSU usecase, the difference being one inherits from UseCase.cc and the other from UseCaseRSU.cc, besides that, all the code is the same.
> > > My goal is for a DEN message sent from a vehicle to be received by the RSU's IcyRoadRSU::indicate(const artery::DenmObject& denm) method.
> > > All is compiling and RSU's are visible in OMNeT window during simulation
> > > ![2019-03-29-161539_1283x646_scrot](https://user-images.githubusercontent.com/15826810/55246855-1547fc80-523e-11e9-930c-aab2e3975d52.png)
> > > But, the RSU's dont seem to be doing anything because IcyRoadRSU::check() is not being invoked for each middleware time step.
> > > The omnet.ini file (https://pastebin.com/77UFGZzz) has _.rsu[_].middleware.updateInterval = 0.1s
> > > Any ideas from where the problem might be coming from?
> > > Best regards and thank you very much for your work and great support of the community with excellent answers!!
> > 
> > 
> > Hi,
> > May I know why did you change the input of `DenService::indicate()` ?
> 
> Hi,
> 
> I don't think I did, as you can see, this is the latest version on artery's git: https://github.com/riebl/artery/blob/master/src/artery/application/DenService.cc
> 
> void DenService::indicate(const vanetza::btp::DataIndication& indication, std::unique_ptrvanetza::UpPacket packet)
> 
> and it is the same

Oh, sorry, I thought you have modified this function and then I found it is the ```indicate`` function in ```DenService``  instead of some usecases. 

Will continue following this question.

**Comment by riebl** [comment]
Does `UseCaseRSU` inherit from `UseCase`? Only implementations of the `den::UseCase` interface are added to the list of use cases in `DenService`, see https://github.com/riebl/artery/blob/26ab50bc57c3b46c429505068029d780013a2a19/src/artery/application/DenService.cc#L64-L66

**Comment by ruipatcheco** [comment]
Hey,

got it all working now, RSU's are able to receive and send messages to Vehicles

The UseCase a class is now inherited by UseCaseVehicle and UseCaseRSU.

I also made SuspendableUseCase :  UseCaseVehicle and SuspendableUseCaseRSU :  UseCaseRSU.


`/*
 * Artery V2X Simulation Framework
 * Copyright 2018 Christina Obermaier
 * Licensed under GPLv2, see COPYING file for detailed license and warranty terms.
 */

#ifndef ARTERY_DEN_USECASE_H_
#define ARTERY_DEN_USECASE_H_

#include <omnetpp/csimplemodule.h>
#include <vanetza/asn1/denm.hpp>
#include <functional>

namespace artery
{

class DenmObject;
class DenService;
class StoryboardSignal;
//class VehicleDataProvider;

namespace den
{
class UseCase : public omnetpp::cSimpleModule
{
public:
    /**
     * Evaluate use case triggering conditions
     * Generates DENM if neccessary
     * \note invoked for each middleware time step (trigger)
     */
    virtual void check() = 0;

    /**
     * Forwards the message received by DenService to the UseCase
     * Messages responding to this can be triggered immediately
     */
    virtual void indicate(const artery::DenmObject&) = 0;

    /**
     * Forwards the received Storyboard trigger signal to the UseCase
     * \note Do not invoke message sending in this method!
     * \note Delay sending until check() is called again.
     * \note This avoids station synchronization.
     */
    virtual void handleStoryboardTrigger(const StoryboardSignal&) = 0;

    void initialize(int);
    int numInitStages () const override { return 1; }

    virtual vanetza::asn1::Denm createMessageSkeleton() = 0;


protected:
    using TriggeringCondition = std::function<bool(void)>;

    DenService* mService = nullptr;

};

} // namespace den
} // namespace artery

#endif /* ARTERY_DEN_USECASE_H_ */
`

`/*
 * Artery V2X Simulation Framework
 * Copyright 2018 Christina Obermaier
 * Licensed under GPLv2, see COPYING file for detailed license and warranty terms.
 */

#ifndef ARTERY_DEN_USECASEVEHICLE_H_
#define ARTERY_DEN_USECASEVEHICLE_H_

#include "artery/application/den/UseCase.h"

namespace artery
{

class VehicleDataProvider;

namespace den
{
class UseCaseVehicle : public UseCase
{
public:
    void initialize(int stage);
    vanetza::asn1::Denm createMessageSkeleton();

protected:
    const VehicleDataProvider* mVdp = nullptr;
};

} // namespace den
} // namespace artery


#endif /* ARTERY_DEN_USECASEVEHICLE_H_ */
`

`#ifndef ARTERY_DEN_SUSPENDABLEUSECASE_H_P2OWWY0L
#define ARTERY_DEN_SUSPENDABLEUSECASE_H_P2OWWY0L

#include "artery/application/den/UseCaseVehicle.h"
#include <boost/optional.hpp>
#include <omnetpp/simtime.h>

namespace artery
{
namespace den
{

class SuspendableUseCase : public UseCaseVehicle
{
public:
    void initialize(int) override;

protected:
    void setDetectionBlockingInterval(omnetpp::SimTime);
    void blockDetection();
    bool isDetectionBlocked();

private:
    omnetpp::SimTime mDetectionBlockingInterval;
    boost::optional<omnetpp::SimTime> mDetectionBlockingSince;
};

} // namespace den
} // namespace artery

#endif /* ARTERY_DEN_SUSPENDABLEUSECASE_H_P2OWWY0L */

`

**Comment by ruipatcheco** [comment]
The problem I'm trying to solve now is how to get RSU's position so I can generate a new DEN message request

vanetza::btp::DataRequestB IcyRoadRSU::createRequest()
{
(...)
    geonet::Area destination;
    geonet::Circle destination_shape;
    destination_shape.r = 100.0 * meter;
    destination.shape = destination_shape;
    
    mService->getFacilities().
    //destination.position.latitude = mVdp->latitude();
    //destination.position.longitude = mVdp->longitude();
    request.gn.destination = destination;

    return request;

**Comment by ruipatcheco** [comment]
Hey! 

I'm now facing what I hope to be my last problem:

to send messages from RSU -> Vehicle, I need a vanetza::btp::DataRequestB where the destination of the message is depicted.

When a vehicle is sending a message, it uses its position (mVdp) as destination  

vanetza::btp::DataRequestB IcyRoad::createRequest()
(...)
    geonet::Area destination;
    geonet::Circle destination_shape;
    destination_shape.r = 100.0 * meter;
    destination.shape = destination_shape;
    destination.position.latitude = mVdp->latitude();
    destination.position.longitude = mVdp->longitude();

But, given RSU's goal is to re-transmit a vehicle's message, it has to set destination as such:

    destination.position.latitude = original Vehicle message . latitude
    destination.position.longitude = original Vehicle message . longitude

The problem comes from how to obtain the original Vehicle message position, since the indicate method only gives us the DENM object and not the request object (which contains the position)

void IcyRoadRSU::indicate(const artery::DenmObject& denm)

So, I've tried to extend DENM Alacarte container to add the position (which is of type GeoAngle):

#include <vanetza/units/angle.hpp>

typedef struct AlacarteContainer {
	LanePosition_t	*lanePosition;	/* OPTIONAL */
	struct ImpactReductionContainer	*impactReduction;	/* OPTIONAL */
	Temperature_t	*externalTemperature;	/* OPTIONAL */
	struct RoadWorksContainerExtended	*roadWorks;	/* OPTIONAL */
	PositioningSolutionType_t	*positioningSolution;	/* OPTIONAL */
	struct StationaryVehicleContainer	*stationaryVehicle;	/* OPTIONAL */
	/*
	 * This type is extensible,
	 * possible extensions are below.
	 */
	vanetza::units::GeoAngle *latitude;
        vanetza::units::GeoAngle *longitude;

But I can't compile it

vanetza/asn1/CMakeFiles/asn1.dir/build.make:998: recipe for target 'vanetza/asn1/CMakeFiles/asn1.dir/its/DecentralizedEnvironmentalNotificationMessage.c.o' failed
In file included from /home/veins/src/artery/extern/vanetza/vanetza/units/angle.hpp:4:0,
                 from /home/veins/src/artery/extern/vanetza/vanetza/asn1/its/AlacarteContainer.h:22,
                 from /home/veins/src/artery/extern/vanetza/vanetza/asn1/its/DecentralizedEnvironmentalNotificationMessage.h:50,
                 from /home/veins/src/artery/extern/vanetza/vanetza/asn1/its/DENM.h:16,
                 from /home/veins/src/artery/extern/vanetza/vanetza/asn1/its/DENM.c:8:
/usr/include/boost/units/absolute.hpp:20:18: fatal error: iosfwd: No such file or directory
 #include <iosfwd>
                                ^
compilation terminated.



So, I tried compiling with another type, the GeoPosition from which GeoAngle stems from:

vanetza::units::GeoAngle longitude() const { return mGeoPosition.longitude; } 

The new struct:

#include "/home/veins/src/artery/src/artery/utility/Geometry.h"

typedef struct AlacarteContainer {
	LanePosition_t	*lanePosition;	/* OPTIONAL */
	struct ImpactReductionContainer	*impactReduction;	/* OPTIONAL */
	Temperature_t	*externalTemperature;	/* OPTIONAL */
	struct RoadWorksContainerExtended	*roadWorks;	/* OPTIONAL */
	PositioningSolutionType_t	*positioningSolution;	/* OPTIONAL */
	struct StationaryVehicleContainer	*stationaryVehicle;	/* OPTIONAL */
	/*
	 * This type is extensible,
	 * possible extensions are below.
	 */
	GeoPosition mGeoPosition;

And it didn't work aswell:

In file included from /usr/include/boost/mpl/bool_fwd.hpp:17:0,
                 from /usr/include/boost/mpl/bool.hpp:17,
                 from /usr/include/boost/mpl/not.hpp:17,
                 from /usr/include/boost/mpl/assert.hpp:17,
                 from /usr/include/boost/geometry/core/closure.hpp:22,
                 from /usr/include/boost/geometry/geometry.hpp:25,
                 from /usr/include/boost/geometry.hpp:17,
                 from /home/veins/src/artery/src/artery/utility/Geometry.h:4,
                 from /home/veins/src/artery/extern/vanetza/vanetza/asn1/its/AlacarteContainer.h:23,
                 from /home/veins/src/artery/extern/vanetza/vanetza/asn1/its/DecentralizedEnvironmentalNotificationMessage.h:50,
                 from /home/veins/src/artery/extern/vanetza/vanetza/asn1/its/DecentralizedEnvironmentalNotificationMessage.c:8:
/usr/include/boost/mpl/aux_/adl_barrier.hpp:33:1: error: unknown type name ‘namespace’

I'm not very experienced in C++ and am a little above my head with these problems... Is this approach correct or is there any other way to achieve the same behaviour?

Thank you for reading, best regards

**Comment by riebl** [comment]
Caller passes `vanetza::btp::DataIndication` object as first argument to `ItsG5BaseService::indicate`. This data indication contains the source position. If you need this information you may need to make it available to use cases as well.
Btw, the compiler error regarding the missing `iosfwd` header occurs because the asn1c generated code is C and iosfwd is only available with C++.

**Comment by ruipatcheco** [comment]
Hello, 

Thanks for your help riebl, I've since been experimenting with your suggested approach.

I got the source position from vanetza::btp::DataIndication object but it is slightly different from the original position, here's an example with latitude:

CAR latitude sent in message = -215072248 
RSU DenService::indicate -> vanetza::btp::DataIndication& indication -> Latitude = -215074024

All RSUs receive the same message that came from the car, the message has a slightly different position as depicted in the example above.

I thought that somewhat during the transport, that packet was being changed due to routing or another factor? So I byte casted the latitude and copied it into the denm alacarte container as such:

char latitude_bytes[sizeof lat];
std::copy(static_cast<const char*>(static_cast<const void*>(&lat)),
              static_cast<const char*>(static_cast<const void*>(&lat)) + sizeof lat,
              latitude_bytes);
msg->denm.alacarte->latitude_bytes = latitude_bytes;

Now here's the tricky part that I can't even make sense of, they are all slightly different:

CAR lat = -215072248
RSU DataIndication = -215074104
RSU alacarte bytes = -215074192


Additionally, the RSUs are receiving the messages but can't send them, they aren't showing any errors, but when setting the destination to the lat/lon of the alacarte bytes, the repetition interval to 1 and maximum to 60 and the destination radius to 1000m they aren't being received by any cars nor other RSUs.

Thanks for reading

**Comment by riebl** [comment]
No packet payload (i.e. messages such as DENM or CAM) are changed during routing. Only some fields such as hop count are modified in GN's basic header. If the message would really change then it is definitely a bug.

The position of `btp::DataIndication` may be different from the one encoded in the message as they are added at different layers, e.g. the position in the indication is added by the router when creating the GBC header. However, there is no delay between creating this header and creating the payload/message; thus position cannot change in between.

If possible consider pushing your modifications to GitHub so I can have a closer look at it and see the whole context.

**Comment by ruipatcheco** [comment]
Hi riebl,

Thanks for your help,

Here is the artery src directory: https://ufile.io/w3qo6ig2

Best regards

**Comment by riebl** [comment]
Dear @ruipatcheco ,

I am sorry but this makes it quite hard to follow your changes. I have not been able to find the reason for your trouble this way.

**Comment by ruipatcheco** [comment]
Completely understandable dear @riebl, 

I'll be sure to put it on github asap, thanks
